enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/EnterpriseConstraintSemantics.java

```diff
@@ -59,48 +61,67 @@ protected PropertyConstraint readNonStandardConstraint( PropertyConstraintRule r
     }
 
     @Override
-    public PropertyConstraintRule writeNodePropertyExistenceConstraint( long ruleId, int type, int propertyKey )
+    public PropertyConstraintRule writeNodePropertyExistenceConstraint( long ruleId, NodePropertyDescriptor descriptor )
     {
-        return nodePropertyExistenceConstraintRule( ruleId, type, propertyKey );
+        return nodePropertyExistenceConstraintRule( ruleId, descriptor );
     }
 
     @Override
-    public PropertyConstraintRule writeRelationshipPropertyExistenceConstraint( long ruleId, int type, int propertyKey )
+    public PropertyConstraintRule writeRelationshipPropertyExistenceConstraint( long ruleId, RelationshipPropertyDescriptor descriptor )
     {
-        return relPropertyExistenceConstraintRule( ruleId, type, propertyKey );
+        return relPropertyExistenceConstraintRule( ruleId, descriptor );
     }
 
     @Override
-    public void validateNodePropertyExistenceConstraint( Iterator<Cursor<NodeItem>> allNodes, int label,
-            int propertyKey ) throws CreateConstraintFailureException
+    public void validateNodePropertyExistenceConstraint( Iterator<Cursor<NodeItem>> allNodes,
+            NodePropertyDescriptor descriptor ) throws CreateConstraintFailureException
     {
         while ( allNodes.hasNext() )
         {
             try ( Cursor<NodeItem> cursor = allNodes.next() )
             {
                 NodeItem node = cursor.get();
-                if ( !node.hasProperty( propertyKey ) )
+                if ( descriptor.isComposite() )
                 {
-                    throw createConstraintFailure( new NodePropertyExistenceConstraintVerificationFailedKernelException(
-                            new NodePropertyExistenceConstraint( label, propertyKey ), node.id() ) );
+                    for ( int propertyKey : descriptor.getPropertyKeyIds() )
```

Comment:
Was just thinking... so many places that now need to do this checking whether or not a descriptor is composite and loop if true, otherwise do single statement (copied code).

Would it be better if the descriptor accepted a visitor instead? Something like:

```
descriptor.visitPropertyKeys( propertyKey -> validateNodePropertyExistenceConstraint( node, propertyKey, descriptor ) );
```